pkg/binder/binder_unit.go

```diff
@@ -267,79 +267,46 @@ func (unitInfo *bindingUnitInfo) GetNewTasksOnNode(nodeName string) []*runningUn
 	return result
 }
 
-func (unitInfo *bindingUnitInfo) GetReadyAndWaitingTasks() []*checkResult {
+func (unitInfo *bindingUnitInfo) getTasks(tasks map[types.UID]*checkResult) []*checkResult {
 	unitInfo.mu.Lock()
 	defer unitInfo.mu.Unlock()
 
-	len := len(unitInfo.readyTasks) + len(unitInfo.waitingTasks)
-	if len == 0 {
+	length := len(tasks)
+	if length == 0 {
 		return nil
 	}
 
-	result := make([]*checkResult, len)
+	result := make([]*checkResult, length)
 	i := 0
-	for _, cr := range unitInfo.readyTasks {
-		result[i] = cr
-		i++
-	}
-	for _, cr := range unitInfo.waitingTasks {
+	for _, cr := range tasks {
 		result[i] = cr
 		i++
 	}
 	return result
 }
 
-func (unitInfo *bindingUnitInfo) GetReadyTasks() []*checkResult {
-	unitInfo.mu.Lock()
-	defer unitInfo.mu.Unlock()
-
-	len := len(unitInfo.readyTasks)
-	if len == 0 {
-		return nil
+func (unitInfo *bindingUnitInfo) GetReadyAndWaitingTasks() []*checkResult {
+	allTasks := make(map[types.UID]*checkResult)
```

Comment:
We should hold the RLock while iterate `readyTasks` & `waitingTasks`, not only inside `getTask`

Reply:
@lzlaa thanks for cr.
These methods will eventually call getTasks and acquire the lock.
```go
func (unitInfo *bindingUnitInfo) getTasks(tasks map[types.UID]*checkResult) []*checkResult {
	unitInfo.mu.Lock()
	defer unitInfo.mu.Unlock()

	length := len(tasks)
	if length == 0 {
		return nil
	}

	result := make([]*checkResult, length)
	i := 0
	for _, cr := range tasks {
		result[i] = cr
		i++
	}
	return result
}

func (unitInfo *bindingUnitInfo) GetReadyAndWaitingTasks() []*checkResult {
	allTasks := make(map[types.UID]*checkResult)
	for uid, cr := range unitInfo.readyTasks {
		allTasks[uid] = cr
	}
	for uid, cr := range unitInfo.waitingTasks {
		allTasks[uid] = cr
	}

	return unitInfo.getTasks(allTasks)
}

func (unitInfo *bindingUnitInfo) GetReadyTasks() []*checkResult {
	return unitInfo.getTasks(unitInfo.readyTasks)
}

func (unitInfo *bindingUnitInfo) GetWaitingTasks() []*checkResult {
	return unitInfo.getTasks(unitInfo.waitingTasks)
}

func (unitInfo *bindingUnitInfo) GetFailedTasks() []*checkResult {
	return unitInfo.getTasks(unitInfo.failedTasks)
}


```

Reply:
I mean we should acquire lock while iterating `unitInfo.readyTasks`, `unitInfo.waitingTasks` before invoking `getTasks` to avoid concurrent conflict caused by updating `unitInfo.readyTasks`, `unitInfo.waitingTasks`.

Reply:
gotcha! thanks for reminding this!

Reply:
done